File: src/main/java/redgear/core/api/tile/IFacedTile.java
package redgear.core.api.tile;

import net.minecraft.entity.EntityLivingBase;
import net.minecraft.item.ItemStack;
import net.minecraft.world.World;
import net.minecraftforge.common.util.ForgeDirection;

public interface IFacedTile {

	/**Returns the forge direction ordinal this block is facing.**/
	int getDirectionId();

	/**Returns the ForgeDirection this tile is facing. **/
	ForgeDirection getDirection();

	/**Sets the facing direction from ForgeDirection ordinal. **/
	boolean setDirection(int id);

	/**
	 * Set the facing direction of this tile
	 * 
	 * @param side Side to set direction to
	 * @return true if block changed, false if it didn't. 
	 */
	boolean setDirection(ForgeDirection side);
	
	/**
	 * Blocks should call this to allow the Tile to decide how to face itself. 
	 * @param world
	 * @param x
	 * @param y
	 * @param z
	 * @param entity
	 * @param stack
	 */
	void onBlockPlacedBy(World world, int x, int y, int z, EntityLivingBase entity, ItemStack stack);
	
}

File: src/main/scala/redgear/core/tile/IFacedTile.scala
package redgear.core.tile

import net.minecraft.entity.EntityLivingBase
import net.minecraft.item.ItemStack
import net.minecraft.util.MathHelper
import net.minecraft.world.World
import net.minecraftforge.common.util.ForgeDirection

/**
 * @author Blackhole
 *         Created on 10/12/2014.
 */
trait IFacedTile {
  //private var _direction: ForgeDirection = ForgeDirection.SOUTH
  //var flat: Boolean = true
  //var allowUnknown: Boolean = false

  def flat: Boolean

  def allowUnknown: Boolean

  def getDirection: ForgeDirection

  def setDirection(side: ForgeDirection)

  /**
   * Blocks should call this to allow the Tile to decide how to face itself.
   * @param world
   * @param x
   * @param y
   * @param z
   * @param entity
   * @param stack
   */
  def onBlockPlacedBy(world: World, x: Int, y: Int, z: Int, entity: EntityLivingBase, stack: ItemStack) = {
    if(flat){
      setDirection(ForgeDirection.getOrientation(IFacedTile.directionMap(MathHelper.floor_double(entity.rotationYaw * 4.0F / 360.0F + 0.5D) & 3)))
    }
    else{
      var quadrant: Int = MathHelper.floor_double(entity.rotationYaw * 4.0F / 360.0F + 0.5D) & 0x3
      quadrant = if (entity.rotationPitch < -60.0F) 5 else if (entity.rotationPitch > 60.0F) 4 else quadrant
      setDirection(ForgeDirection.getOrientation(IFacedTile.directionMap(quadrant)))
    }


  }
}

object IFacedTile{
  private val directionMap: Array[Int] = Array(2, 5, 3, 4, 1, 0)

   def setDirection(self: IFacedTile, side: ForgeDirection) = {

      val fn = {if (!(self flat)) self setDirection side}
      side match {
        case ForgeDirection.UNKNOWN => {if (self allowUnknown) self setDirection side}
        case ForgeDirection.UP => fn
        case ForgeDirection.DOWN => fn
        case _ => self setDirection side
      }
    }
}
